File: internal/probe/tcp/tcp.go
package tcp

import (
	"context"
	"net"
	"time"

	"icurl/internal/evidence"
	"icurl/internal/probe"
)

type Probe struct {
	Timeout time.Duration
}

func (p Probe) Run(ctx context.Context, target probe.Target) evidence.ProbeResult {
	started := time.Now()
	address := net.JoinHostPort(target.Host, target.Port)
	result := evidence.ProbeResult{
		ProbeName:  "TCP",
		Layer:      evidence.LayerTCP,
		Target:     address,
		StartedAt:  &started,
		Confidence: evidence.ConfidenceObserved,
	}

	timeout := p.Timeout
	if timeout == 0 {
		timeout = 5 * time.Second
	}
	dialer := net.Dialer{Timeout: timeout}
	
	var conn net.Conn
	var err error
	if len(target.IPs) > 0 {
		for _, ip := range target.IPs {
			addr := net.JoinHostPort(ip.String(), target.Port)
			conn, err = dialer.DialContext(ctx, "tcp", addr)
			if err == nil {
				break
			}
		}
	} else {
		conn, err = dialer.DialContext(ctx, "tcp", address)
	}

	if err != nil {
		kind := evidence.ErrorUnknown
		if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
			kind = evidence.ErrorTimeout
		}
		finish(&result, evidence.ResultFailed, kind, err.Error())
		return result
	}
	defer func() {
		_ = conn.Close()
	}()

	result.RemoteAddress = conn.RemoteAddr().String()
	result.LocalAddress = conn.LocalAddr().String()
	result.Observations = []string{"TCP connect succeeded"}
	finish(&result, evidence.ResultOK, evidence.ErrorNone, "")
	return result
}

func finish(result *evidence.ProbeResult, probeResult evidence.Result, kind evidence.ErrorKind, message string) {
	finished := time.Now()
	result.FinishedAt = &finished
	result.DurationNS = finished.Sub(*result.StartedAt).Nanoseconds()
	result.Result = probeResult
	result.ErrorKind = kind
	result.ErrorMessage = message
}

File: icurl-rs/src/probe/tcp.rs
use std::time::Duration;
use crate::evidence::{ErrorKind, Layer, ProbeResult, ProbeResultStatus};
use crate::probe::{BoxFuture, Probe, Target};
use crate::probe::tcp_connect::{try_connect, record_addresses};

pub struct TcpProbe {
    pub timeout: Duration,
}

impl TcpProbe {
    pub fn new(timeout: Duration) -> Self {
        Self { timeout }
    }
}

impl Probe for TcpProbe {
    fn probe_name(&self) -> &'static str { "TCP" }

    fn run<'a>(&'a self, target: &'a Target) -> BoxFuture<'a, ProbeResult> {
        Box::pin(async move {
            let address = format!("{}:{}", target.host, target.port);
            let mut result = ProbeResult::new("TCP", Layer::Tcp, &address);

            let timeout_duration = if self.timeout.is_zero() {
                Duration::from_secs(5)
            } else {
                self.timeout
            };

            match try_connect(&target.ips, &target.host, target.port, timeout_duration).await {
                Ok(stream) => {
                    record_addresses(&stream, &mut result);
                    result.add_observation("TCP connect succeeded");
                    result.finish(ProbeResultStatus::Ok, ErrorKind::None, "");
                }
                Err((kind, msg)) => {
                    result.finish(ProbeResultStatus::Failed, kind, &msg);
                }
            }

            result
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tokio::net::TcpListener;

    #[tokio::test]
    async fn test_tcp_probe_success() {
        let listener = TcpListener::bind("127.0.0.1:0").await.unwrap();
        let addr = listener.local_addr().unwrap();

        tokio::spawn(async move {
            let _ = listener.accept().await;
        });

        let target = Target {
            url: url::Url::parse("https://example.com").unwrap(),
            host: addr.ip().to_string(),
            port: addr.port(),
            ips: vec![addr.ip()],
        };

        let probe = TcpProbe::new(Duration::from_secs(1));
        let res = probe.run(&target).await;
        assert_eq!(res.result, ProbeResultStatus::Ok);
        assert_eq!(res.error_kind, ErrorKind::None);
        assert!(res.remote_address.is_some());
    }
}
